### PR TITLE
[STORM-3692] Handle UID return from Files.getOwner()

### PR DESCRIPTION
**Pull request by bipinprasad**
## What is the purpose of the change

*(Pushback to Community) Owner on /proc/pid is sometimes returned as UID instead of username in production. Handle this special case.*

## How was the change tested

*New test was written to mock this condition and test the correct behavior of ServerUtils.isAnyPosixProcessPidDirAlive()*

### CODE REVIEW

File: storm-server/src/main/java/org/apache/storm/utils/ServerUtils.java [hunk 1]
```diff
@@ -1217,9 +1199,127 @@ public static boolean areAllProcessesDead(Map<String, Object> conf, String user,
...
+                        throw new IOException(err);
+                    }
+                    cachedUserToUidMap.put(user, uid);
+                }
+                if (uid == pidUid) {
+                    LOG.debug("Process {} is alive and owned by user {}/{}", pid, user, uid);
+                    return true;
+                } else {
+                    LOG.info("Prior process is dead, since directory {} owner {} is not same as expected user {}/{}, "
+                            + "likely pid {} was reused for a new process for uid {}", pidDir, pidUser, user, uid, pid, uid);
```
Comment (Ethanlm) [inline storm-server/src/main/java/org/apache/storm/utils/ServerUtils.java:1306]: Should the last "uid" be "pidUid"?

It is not easy to distinguish "user" vs "pidUser", "uid" vs "pidUid". I suggest to change them to be 
`actualUser vs expectedUser, actualUid vs expectedUid`

File: storm-server/src/test/java/org/apache/storm/utils/ServerUtilsTest.java [hunk 2]
```diff
@@ -162,6 +185,191 @@ public void testGetUserId() throws Exception {
...
         Path p = Files.createTempFile("testGetUser", ".txt");
         int uid2 = ServerUtils.getPathOwnerUid(p.toString());
+        if (!p.toFile().delete()) {
+            LOG.warn("Could not delete tempoary file {}", p);
+        }
         assertEquals("User UID " + uid1 + " is not same as file " + p.toString() + " owner UID of " + uid2, uid1, uid2);
     }
+
+    @Test
+    public void testIsAnyProcessPosixProcessPidDirAlive() throws IOException {
```
Comment (Ethanlm) [inline storm-server/src/test/java/org/apache/storm/utils/ServerUtilsTest.java:195]: Isn't this already present? 
https://github.com/apache/storm/blob/master/storm-server/src/test/java/org/apache/storm/utils/ServerUtilsTest.java#L184

File: storm-server/src/test/java/org/apache/storm/utils/ServerUtilsTest.java [hunk 3]
```diff
@@ -162,6 +185,191 @@ public void testGetUserId() throws Exception {
...
+        }
+    }
+
+    /**
+     * Sleep for specified milliseconds and return true if sleep was interrupted.
+     *
+     * @param milliSeconds number of milliseconds to sleep
+     * @return true if sleep was interrupted, false otherwise.
+     */
+    private boolean sleepInterrupted(long milliSeconds) {
```
Comment (Ethanlm) [inline storm-server/src/test/java/org/apache/storm/utils/ServerUtilsTest.java:365]: this method already exist?

File: storm-server/src/test/java/org/apache/storm/utils/ServerUtilsTest.java [hunk 4]
```diff
@@ -162,6 +185,191 @@ public void testGetUserId() throws Exception {
...
+    }
+
+    /**
+     * Make the best effort to obtain the Process ID from the Process object. Thus staying entirely with the JVM.
+     *
+     * @param p Process instance returned upon executing {@link Runtime#exec(String)}.
+     * @param errors Populate errors when PID is a negative number.
+     * @return positive PID upon success, otherwise negative.
+     */
+    private synchronized long getPidOfPosixProcess(Process p, List<String> errors) {
```
Comment (Ethanlm) [inline storm-server/src/test/java/org/apache/storm/utils/ServerUtilsTest.java:307]: this method already exist?

File: storm-server/src/test/java/org/apache/storm/utils/ServerUtilsTest.java [hunk 5]
```diff
@@ -162,6 +185,191 @@ public void testGetUserId() throws Exception {
...
+                    LOG.info("{}: ({}) Found no existing process directories after killing last process", testName, i);
+                }
+            }
+        }
+        if (!errors.isEmpty()) {
+            fail(String.format("There are %d failures in test:\n\t%s", errors.size(), String.join("\n\t", errors)));
+        }
+    }
+
+
```
Comment (Ethanlm) [inline storm-server/src/test/java/org/apache/storm/utils/ServerUtilsTest.java:270]: nit: more than 1 new line

File: storm-server/src/test/java/org/apache/storm/utils/ServerUtilsTest.java [hunk 6]
```diff
@@ -162,6 +185,191 @@ public void testGetUserId() throws Exception {
...
+        File procDir = new File("/proc");
+        if (!procDir.exists()) {
+            LOG.info("Test testIsAnyPosixProcessPidDirAlive is designed to run on systems with /proc directory only, marking as success");
+            return;
+        }
+        Collection<Long> pids = getRunningProcessIds();
+        assertFalse(pids.isEmpty());
+
+        for (int i = 0 ; i < 2 ; i++) {
+            boolean mockFileOwnerToUid = (i % 2 == 0);
```
Comment (Ethanlm) [inline storm-server/src/test/java/org/apache/storm/utils/ServerUtilsTest.java:290]: I would suggest to use Parameterized Test here.

File: storm-server/src/main/java/org/apache/storm/utils/ServerUtils.java (hunk 1 above)
Comment (bipinprasad) [reply storm-server/src/main/java/org/apache/storm/utils/ServerUtils.java:1306]: renamed variables

File: storm-server/src/test/java/org/apache/storm/utils/ServerUtilsTest.java (hunk 5 above)
Comment (bipinprasad) [reply storm-server/src/test/java/org/apache/storm/utils/ServerUtilsTest.java:270]: removed

File: storm-server/src/test/java/org/apache/storm/utils/ServerUtilsTest.java (hunk 4 above)
Comment (bipinprasad) [reply storm-server/src/test/java/org/apache/storm/utils/ServerUtilsTest.java:307]: I am not aware of any existing method with similar functionality.

File: storm-server/src/test/java/org/apache/storm/utils/ServerUtilsTest.java (hunk 3 above)
Comment (bipinprasad) [reply storm-server/src/test/java/org/apache/storm/utils/ServerUtilsTest.java:365]: I am not aware of any existing method with similar functionality.

File: storm-server/src/test/java/org/apache/storm/utils/ServerUtilsTest.java (hunk 2 above)
Comment (bipinprasad) [reply storm-server/src/test/java/org/apache/storm/utils/ServerUtilsTest.java:195]: Yes. Something is not right. I will do a force push with all changes.

File: storm-server/src/test/java/org/apache/storm/utils/ServerUtilsTest.java (hunk 3 above)
Comment (Ethanlm) [reply storm-server/src/test/java/org/apache/storm/utils/ServerUtilsTest.java:365]: https://github.com/apache/storm/blob/master/storm-server/src/test/java/org/apache/storm/utils/ServerUtilsTest.java#L324
This method already exists in ServerUtilsTest (from your another merged PR)

File: storm-server/src/test/java/org/apache/storm/utils/ServerUtilsTest.java (hunk 4 above)
Comment (Ethanlm) [reply storm-server/src/test/java/org/apache/storm/utils/ServerUtilsTest.java:307]: https://github.com/apache/storm/blob/master/storm-server/src/test/java/org/apache/storm/utils/ServerUtilsTest.java#L266
This method already exists in ServerUtilsTest (from your another merged PR)

File: storm-server/src/test/java/org/apache/storm/utils/ServerUtilsTest.java (hunk 2 above)
Comment (Ethanlm) [reply storm-server/src/test/java/org/apache/storm/utils/ServerUtilsTest.java:195]: You need to rebase on latest master. Then you should be able to see the conflicts/duplicates

**Pull request merged**